src-tauri/src/terminal.rs

```diff
@@ -0,0 +1,414 @@
+                cmd.args([
+                    "-e",
+                    "/usr/bin/env",
+                    "sh",
+                    "-c",
+                ]).args(command);
+                cmd
+            }
+            Terminal::Terminal => {
+                // Create a bash script and run that. Not very secure but it makes this easier.
+                let download_script = format!("#!/bin/bash\ncd {}\n{}",env::current_dir().unwrap().display(), command[0]);
+                // println!("{}", download_script);
+
+                fs::write("./script.sh", download_script).unwrap();
+                fs::set_permissions("./script.sh", fs::Permissions::from_mode(0o755)).unwrap(); // Won't run without executable permission
+
+                let mut cmd = Command::new("/usr/bin/open");
+                cmd.args(&["-a", "Terminal", "./script.sh"]);
+                cmd
+            }
+        }
+    }
+    //endregion
+}
+
+/**
+Checks if terminals are installed by checking if they respond to commands.
+
+## How it works
+Probes a list of popular terminals and checks if they return an error when calling their `--version` or similar command line flag.
+
+## Options
+* `return_immediately`: [`bool`]: Return as soon as one terminal is found.
+
+## Returns
+A vector containing a list of terminals that should work.
+
+## Commands
+| Terminal       | Command to check if installed |
+|----------------|-------------------------------|
+| cmd            | `cmd /?`                      |
+| GNOMETerminal  | `gnome-terminal --version`    |
+| GNOMEConsole   | `kgx --version`               |
+| Konsole        | `konsole --version`           |
+| Xfce4Terminal  | `xfce4-terminal --version`    |
+| Terminator     | `terminator --version`        |
+| Terminology    | `terminology --version`       |
+| XTerm          | `xterm -v`                    |
+| Kitty          | `kitty --version`             |
+| LXTerminal     | `lxterminal --version`        |
+| Tilix          | `tilix --version`             |
+| DeepinTerminal | `deepin-terminal --version`   |
+| CoolRetroTerm  | `cool-retro-term --version`   |
+| Alacritty      | `alacritty --version`         |
+
+ */
+pub async fn get_installed_terminals(return_immediately: bool) -> Vec<Terminal> {
+    match get_os() {
+        "windows" => { return vec!(Terminal::CMD); }
+        "macos" => { return vec!(Terminal::Terminal); }
+        _ => {}
+    }
+
+
+    let mut available_terminals: Vec<Terminal> = Vec::new();
+
+    for terminal in Terminal::iter() {
+        // Probe terminal. If it doesn't raise an error, it is probably installed.
+        if terminal.installed().await {
+            if return_immediately {
+                return vec![terminal];
+            }
+            available_terminals.push(terminal);
+        }
+    }
+
+    if available_terminals.is_empty() {
+        eprintln!("No terminals were detected. Try installing one.");
+    }
+
+    available_terminals
```

Comment:
Is this some rust thing that the return keyword isn't needed?

Reply:
Yes!

Reply:
strange